src/lua/init.lua

```diff
@@ -86,81 +111,75 @@ local function load_lua(file)
-            strerr = strerr .. err
-        end
-        index = string.find(cwd, "/[^/]*$")
+    while path ~= '/' do
+        path = fio.dirname(path)
+        table.insert(paths, path)
     end
-    return nil, strerr
-end
 
-local function search_rocks_lua(name)
-    local pathes_search = {
-        "/.rocks/share/tarantool/?.lua;",
-        "/.rocks/share/tarantool/?/init.lua;",
-    }
-    return traverse_rocks(name, pathes_search)
-end
+    return paths
+end
+
+-- Generate a search function, which performs searching through
+-- templates setup in options.
+--
+-- @param opts table with following parameters:
+--   - path_fn function which returns a base path for the
+--     resulting template
+--   - templates table with lua search templates
+--   - traverse bool flag which tells search function to build
+--     multiple paths by expanding base path up to the root ('/')
+-- @return a searcher function which builds a path template and
+-- calls package.searchpath
+local function gen_search_func(opts)
+    assert(type(opts) == 'table', 'opts must be a table')
+    assert(type(opts.path_fn) == 'function', 'opts.path_fn must be a function')
+    assert(type(opts.templates) == 'table', 'opts.templates must be a table')
 
-local function search_rocks_lib(name)
-    local pathes_search = {
-        "/.rocks/lib/tarantool/?."..soext
-    }
-    return traverse_rocks(name, pathes_search)
-end
-
-local function cwd_loader_func(lib)
-    local search_cwd = lib and search_cwd_lib or search_cwd_lua
-    local load_func = lib and load_lib or load_lua
     return function(name)
-        if not name then
-            return "empty name of module"
-        end
-        local file, err = search_cwd(name)
-        if not file then
-            return err
-        end
-        local loaded, err = load_func(file, name)
-        if err == nil then
-            return loaded
-        else
-            return err
+        local path = opts.path_fn() or '.'
+        local paths = opts.traverse and traverse_path(path) or { path }
+
+        local searchpaths = {}
+
+        for _, path in ipairs(paths) do
+            for _, template in pairs(opts.templates) do
+                table.insert(searchpaths, fio.pathjoin(path, template))
+            end
         end
+
+        local searchpath = table.concat(searchpaths, ';')
+
+        return package.searchpath(name, searchpath)
     end
 end
 
-local function rocks_loader_func(lib)
-    local search_rocks = lib and search_rocks_lib or search_rocks_lua
-    local load_func = lib and load_lib or load_lua
-    return function (name)
+-- Compose a loader function from options.
+--
+-- @param opts table with the following keys:
+--   - search_fn function will be used to search a file from path
+--     template
+--   - load_fn function will be used to load a file, found by
+--     search function
+-- @return function a loader, which first search for the file and
+-- then loads it
+local function gen_loader_func(opts)
```

Comment:
The comment as about `gen_search_func`.

Reply:
Switched to positional arguments here too